pint/formatting.py

```diff
@@ -117,6 +117,135 @@ def _pretty_fmt_exponent(num):
     },
 }
 
+#: _FORMATTERS maps format names to callables doing the formatting
+_FORMATTERS: Dict[str, Callable] = {}
+
+
+def register_unit_format(name):
+    """register a function as a new format for units
+
+    The registered function must have a signature of:
+
+    .. code:: python
+
+        def new_format(unit, registry, **options):
+            pass
+
+    Parameters
+    ----------
+    name : str
+        The name of the new format (to be used in the format mini-language). A error is
+        raised if the new format would overwrite a existing format.
+
+    Examples
+    --------
+    >>> @pint.register_unit_format("custom")
+    ... def format_custom(unit, registry, **options):
+    ...     result = "<formatted unit>"  # do the formatting
+    ...     return result
```

Comment:
this is executed twice (once for `pint.formatting.register_unit_format` and once for `pint.register_unit_format`), and because there's no isolation in between the runs, the second run will try to register a pre-existing name. Not sure how to avoid that... does `sphinx` have a test isolation feature, or should we switch to `python -m pytest --doctest-modules` and the `ipython` / `jupyter` `sphinx` directives?

Reply:
as far as I can tell, there's not much we can do here, except switching the doctest runner to `pytest` (and the `doctest` blocks in the `rst` files to `ipython` / `jupyter-execute`), to replace `autodoc` with `autosummary`, or to make sure `autodoc` does not generate pages for both `pint.register_unit_format` and `pint.formatting.register_unit_format`.

Since all of these are out-of-scope for this PR, I'll convert the example to a code block and we can discuss further changes in a dedicated issue / PR.

Reply:
Yes, there is no way out.